### PR TITLE
Csiro/s3 driver

### PR DESCRIPTION
**Pull request by petewa**
### Reason for this pull request

Improvements to DataCube:
   - S3 storage backend
      - Windows is not supported.
   - driver manager to dynamically load/switch storage drivers (NetCDF, S3, S3-test).
   - Ingest now supports creation of nD Storage Units in the available storage drivers.
   	 e.g.
   	 	NetCDF:  datacube -v ingest -c ls5_nbar_albers.yaml --executor multiproc 8
		S3:      datacube --driver s3 -v ingest -c  ls5_nbar_albers_s3.yaml --executor multiproc 8
		S3-test: datacube --driver s3-test -v ingest -c ls5_nbar_albers_s3_test.yaml --executor multiproc 8
  - load and load_data has optional multi-threaded support with use_threads flag.

Improved testing:
 - tests are run for each driver, where possible.
 - tests corner values
 - tests md5 hash equality on load_data
 - tests multiple time slices
 - reduction in data usage
 - reduction in total number of concurrent db connections.

### Proposed changes
 - cli driver parameter to select driver, if None, it defaults to NetCDF.
 - support for generating n-dimension storage units on ingest.
   - example ingest yaml: docs/config_samples/ingester/ls5_nbar_nbar_s3.yaml
 - supported drivers:
   - NetCDF: based on existing driver.
   - S3: S3 backend for storage.
   - S3-test: Same as S3 but emulated on disk.
 - datacube.api.load_data - use_threads parameter to enable threaded _fuse_measurement with results stored in a shared memory array.
 - datacube.scripts.ingest - uses slightly modified GridWorkFlow to generate 3D tasks for 3D Storage Unit creation.
 - optional creation of s3 tables via "datacube -v system init -s3"

### Todo:
 - More tests.

 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [ ] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/drivers/index.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+          :class:`datacube.index._api.Index` and used for testing
+          purposes only. In the current implementation, only the
+          `index._db` variable is used, and is passed to the index
+          initialisation method, that should basically replace the
+          existing DB connection with that variable.
+        :param args: Optional positional arguments to be passed to the
+          index on initialisation. Caution: In the current
+          implementation all parameters get passed to all potential
+          indexes.
+        :param args: Optional keyword arguments to be passed to the
```
Comment (Kirill888) [inline datacube/drivers/index.py]: should be `:params kargs:` (missing `k`)

File: datacube/drivers/driver.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        if not self.__index:
+            raise ValueError('Driver was not initialised properly. Make sure you call the base class _init__')
+        return self.__index
+
+    def as_uri(self, path):
+        """Set or replace the uri scheme for a path according to the driver's.
+        """
+        path = str(path)
+        body = path.split(':', 1)[1] if ':' in path else path
+        return '%s:%s' % (self.uri_scheme, body)
```
Comment (omad) [inline datacube/drivers/driver.py]: This is overly simplistic URI generation, in particular, local file URLs should start like `file:///myrootfile.txt` not `file:/myrootfile.txt'. There is code elsewhere that handles this, we'll just need to plumb it through.

File: datacube/drivers/manager.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
+"""Module used to dynamically load storage drivers.
+"""
+from __future__ import absolute_import
+
+import logging
+import sys
+import weakref
+from pathlib import Path
+from collections import Iterable
+from dill import loads, dumps
```
Comment (omad) [inline datacube/drivers/manager.py]: It would be nice to test whether `cloudpickle` will work instead of `dill`. I imagine it does, and we already depend on it via `xarray/dask`.

File: datacube/drivers/manager.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+from .driver import Driver
+from .index import Index
+
+# Dynamic loading from filename varies across python versions
+# Based on http://stackoverflow.com/a/67692
+if sys.version_info >= (3, 5):  # python 3.5+
```
Comment (omad) [inline datacube/drivers/manager.py]: I love that there's support for multiple python versions! We normally try and keep all of this code in a central `datacube/compat.py` file and import required functions from there.

File: datacube/drivers/manager.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    dataset, in particular, without knowing what driver is required to
+    retrieve the actual data.
+
+    A 'current' driver can be set in the manager and which can be used
+    to handle datasets which don't specifiy a storage driver in their
+    metadata.
+    """
+
+    _DEFAULT_DRIVER = 'NetCDF CF'
+    '''Default 'current' driver, assuming its code is present.
```
Comment (omad) [inline datacube/drivers/manager.py]: Instead of 'docstrings' for class attributes, the convention is to use
```python
    #: Default 'current' driver, assuming its code is present.
    _DEFAULT_DRIVER = 'NetCDF CF'
```

File: datacube/index/postgres/tables/_schema.py [hunk 6]
```diff
@@ -111,3 +112,92 @@
     PrimaryKeyConstraint('dataset_ref', 'classifier'),
     UniqueConstraint('source_dataset_ref', 'dataset_ref'),
 )
+
+
+# --- S3-driver specific Tables ---
+S3_DATASET = Table(
```
Comment (omad) [inline datacube/index/postgres/tables/_schema.py:118]: I had expected the S3 support to be optional, ie. only create new database tables if it's enabled.

It's not a huge problem having extra tables always being created, but optional would have been nicer.

File: datacube/drivers/manager.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+          indexes.
+        """
+        if self.__drivers:
+            for driver in self.__drivers.values():
+                # pylint: disable=protected-access
+                if driver.index._db != self.__index._db:
+                    driver.index.close()
+        self.__drivers = {}
+        for init_path in Path(__file__).parent.glob('*/__init__.py'):
+            init = load_module(str(init_path.parent.stem), str(init_path))
```
Comment (omad) [inline datacube/drivers/manager.py:160]: Around here could deal with `ImportErrors` with drivers, when they depend on packages which aren't installed. Drivers could also check that any other pre-requirements are dealt with, ie. if necessary Database tables were available for use, and fail silently with a warning if not.

File: datacube/index/postgres/_api.py [hunk 8]
```diff
@@ -868,3 +873,124 @@ def grant_role(self, role, users):
                 raise ValueError('Unknown user %r' % user)
 
         tables.grant_role(self._connection, pg_role, users)
+
+
+    # S3 specific functions
+    # See .tables for description of each column
+    def put_s3_mapping(self, dataset_ref, band, s3_dataset_id):
```
Comment (omad) [inline datacube/index/postgres/_api.py]: Still not convinced by s3 specifics polluting this already crowded class. It would be great to have these separate. If we get more drivers the problem will only become worse.

File: datacube/api/core.py [hunk 9]
```diff
@@ -1,21 +1,26 @@
...
 from collections import namedtuple, OrderedDict
 from math import ceil
 import warnings
 
+import os
+import uuid
 import pandas
 import numpy
 import xarray
+import SharedArray as sa
```
Comment (omad) [inline datacube/api/core.py]: I'd really like `SharedArray` to be an optional dependency. It's currently only available as a source package requiring compilation, which makes installing it much more difficult on Windows/Mac systems without a compiler.

It's also slowed down the review of all this work, since none of the existing environments available to me could run this branch.

**Review by omad — CHANGES_REQUESTED** [review]
Looks great guys, Love all the documentation and testing, looks like high quality code.

There are however a couple of issues, a couple of which I consider blockers:
- [ ] Mandatory dependency on `SharedArray`
- [ ] Slightly erroneous URI handling, which is the cause of the current test failures.

And a few minor things which shouldn't be hard to correct:
- [ ] Use `cloudpickle` instead of `dill`, since we already depend on it and afaik they do the same job.
- [ ] Move multi-version python compatible code into the `compat` module.

**Comment by omad** [comment]
Thanks @petewa , this is looking much closer to what we both had in mind.

I've gone further towards making the S3 dependencies optional by making the S3 specific tests `skip` if they can't be run.

Still outstanding is support for Windows filenames in the Driver Manager. When running on Windows there is still a failing few tests:
```
_______________ test_index_dataset_with_location[NetCDF CF-UTC] _______________

index = Index<db=PostgresDb<engine=Engine(postgresql://localhost:5432/agdcintegration)>>
default_metadata_type = MetadataType(name='eo', id_=1)
driver = <NetCDFDriver.NetCDFDriver object at 0x000000000F15DE48>

    def test_index_dataset_with_location(index, default_metadata_type, driver):
        """
        :type index: datacube.index._api.Index
        :type default_metadata_type: datacube.model.MetadataType
        """
        first_file = Path('/tmp/first/something.yaml').absolute()
        second_file = Path('/tmp/second/something.yaml').absolute()
        first_uri = driver.as_uri(first_file)
        second_uri = driver.as_uri(second_file)

        type_ = index.products.add_document(_pseudo_telemetry_dataset_type)
        dataset = Dataset(type_, _telemetry_dataset, uris=[first_uri], sources={})
        index.datasets.add(dataset)
        stored = index.datasets.get(dataset.id)
        assert stored.id == _telemetry_uuid
        # TODO: Dataset types?
        assert stored.type.id == type_.id
        assert stored.metadata_type.id == default_metadata_type.id
        if driver.uri_scheme == 'file':
>           assert stored.local_path == Path(first_file)
E           AssertionError: assert WindowsPath('.') == WindowsPath('C:/tmp/first/something.yaml')
E            +  where WindowsPath('.') = Dataset <id=4ec8fe97-e8b9-11e4-87ff-1040f381a756 type=ls8_telemetry location=.>.local_path
E            +  and   WindowsPath('C:/tmp/first/something.yaml') = Path(WindowsPath('C:/tmp/first/something.yaml'))
```

**Comment by omad** [comment]
Hi @petewa and everyone. I'm doing a last look over everything this afternoon, and am hoping to merge it in then. I want to work out why the test coverage has dropped, maybe some tests have been skipped.

Watch this space.

**Review by omad — APPROVED** [review]
